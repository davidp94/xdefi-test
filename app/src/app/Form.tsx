import React, { useState } from "react";
import { useForm } from "react-hook-form";
import styled from "styled-components";
import { Button } from "../components/Button";
import { TextInput } from "../components/TextInput";
import { Signature } from "./Signature";

const Wrapper = styled.div`
  width: 700px;
  display: flex;
  flex: 0 0 auto;
`;

const Content = styled.div`
  display: flex;
  flex: 1 1 auto;
  flex-direction: column;
`;

const Row = styled.div`
  display: flex;
  flex-direction: row;
  flex: 1 1 auto;
`;

const PrivateKeyTextInput = styled(TextInput)`
  margin: 0px 5px 10px 0px;
`;

const MessageTextInput = styled(TextInput)`
  margin: 0px 0px 10px 5px;
`;

export const Form = () => {
  const [signature, setSignature] = useState<String>("");
  const [resultVerify, setResultVerify] = useState<String>("");
  const [pubkey, setPubkey] = useState<String>("");

  const {
    register,
    handleSubmit,
    formState: { isDirty },
  } = useForm();

  const onSubmit = handleSubmit(async (data) => {
    const promise = await import("elliptic-curve");
    const result = promise.sign(data.privateKey, data.message);
    const resultPubkey = promise.get_pubkey(data.privateKey);

    const resultVerify = promise.verify(resultPubkey, data.message, result);

    setPubkey(resultPubkey);
    setSignature(result);
    setResultVerify(resultVerify);
  });

  return (
    <form onSubmit={onSubmit}>
      <Wrapper>
        <Content>
          <Row>
            <PrivateKeyTextInput
              {...register("privateKey")}
              placeholder='Please enter your private key here'
              testId='private-key'
            />
            <MessageTextInput
              {...register("message")}
              placeholder='Please enter your message here'
              testId='message'
            />
          </Row>
          <Signature>{signature}</Signature>
          <Row>
            <Button onClick={onSubmit} testId='submit' disabled={!isDirty}>
              Sign
            </Button>
          </Row>
          <div>
            Result pubkey: <br />
            {pubkey}
          </div>
          <div>
            Result verify: <br />
            {resultVerify}
          </div>
        </Content>
      </Wrapper>
    </form>
  );
};
