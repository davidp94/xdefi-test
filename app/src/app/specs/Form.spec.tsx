import React from "react";
import { fireEvent, render } from "@testing-library/react";
import { Form } from "../Form";

describe("<Form />", () => {
  beforeEach(() => {
    jest.mock("elliptic-curve", () => {
      return jest.fn().mockImplementation(() => ({
        sign: () => "SIGNATURE",
      }));
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("should render", () => {
    const component = render(<Form />);
    expect(component).toMatchSnapshot();
  });

  it("should generate signature", () => {
    const { getByTestId } = render(<Form />);

    const privateKeyField = getByTestId("private-key");
    const messageField = getByTestId("message");
    const submitButton = getByTestId("submit");
    const signature = getByTestId("signature");

    fireEvent.change(
      privateKeyField,
      "03c1e9550e66958296d11b60f8e8e7a7ad990d07fa65d5f7652c4a6c87d4e3cc"
    );
    fireEvent.change(messageField, "Hello World");

    fireEvent.click(submitButton);

    expect(signature.textContent).toEqual("SIGNATURE");
  });
});
