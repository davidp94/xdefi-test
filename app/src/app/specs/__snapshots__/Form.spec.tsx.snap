// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="sc-hKwCoD lofpdH"
        >
          <div
            class="sc-eCImvq lfOzRy"
          >
            <div
              class="sc-jRQAMF bWtpao"
            >
              <textarea
                class="sc-gsDJrp dYPcJk sc-gKckTs kYgzSu"
                data-testid="private-key"
                name="privateKey"
                placeholder="Please enter your private key here"
              />
              <textarea
                class="sc-gsDJrp dYPcJk sc-iCfLBT eXmbVL"
                data-testid="message"
                name="message"
                placeholder="Please enter your message here"
              />
            </div>
            <div
              class="sc-jRQAMF bWtpao"
            >
              <div
                class="sc-bdvvaa jcZsnB"
              >
                Sign
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="sc-hKwCoD lofpdH"
      >
        <div
          class="sc-eCImvq lfOzRy"
        >
          <div
            class="sc-jRQAMF bWtpao"
          >
            <textarea
              class="sc-gsDJrp dYPcJk sc-gKckTs kYgzSu"
              data-testid="private-key"
              name="privateKey"
              placeholder="Please enter your private key here"
            />
            <textarea
              class="sc-gsDJrp dYPcJk sc-iCfLBT eXmbVL"
              data-testid="message"
              name="message"
              placeholder="Please enter your message here"
            />
          </div>
          <div
            class="sc-jRQAMF bWtpao"
          >
            <div
              class="sc-bdvvaa jcZsnB"
            >
              Sign
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
